### DISCUSSION TITLE
How to avoid tick duplicates when using `tickFormatter`?

### DISCUSSION THREAD

**Discussion by dmdum** [recharts/recharts#4344] [Q&A] [answered]
I'm using `tickFormatter` to to change how the ticks are displayed. I would like it so there are no duplicates after the ticks are transformed. Using `allowDuplicatedCategory: false` doesn't seem to work for `tickFormatter`. 

Is there any other way I can avoid duplicates?
Thank you.

**Comment by ckifer** [comment, marked as answer]
allowDuplicatedCategory only respects the actual values given to the chart. You will have to dedupe yourself in that case

**Reply by dmdum** [reply]
I see, thank you!

**Comment by shettyvega** [comment]
@ckifer I have a similar problem. My datakey is a timestamp but I am formatting it to be MM'YYYY, Because of this, I have repeating months show up on the x axis. Is there a getTicks function I can use and filter out the repeating ticks myself? Thanks.

**Reply by ckifer** [reply]
you can specify values you want to show in the `ticks` prop. 

Does this help?
https://github.com/recharts/recharts/issues/4537#issuecomment-2118393596
https://codesandbox.io/p/sandbox/recharts-issue-template-forked-4qpg3k

**Reply by ckifer** [reply]
In your case you would only add a value to your ticks array if you don't already have one for that month

**Reply by shettyvega** [reply]
So basically I will have to create the ticks array with the evenly distributed timestamps (without month repetition) and the tickFormatter (MM'YYYY) remains as is.

**Reply by ckifer** [reply]
Correct. Don't worry about distribution until you see what the chart looks like after crafting the array

Psuedo code:
```
// assume the array is sorted by date ascending

getTicks(data)
  ticks = []
  for item of data
    month = getMonthOfItem(item)
    if ticks does not have an entry with `month`
      ticks.push(month)
   return ticks
      
      
`XAxis ticks={ticks} ... />`
```

**Reply by shettyvega** [reply]
I will give this a try. Thank you for the prompt replies.

**Reply by shettyvega** [reply]
This is what I have. However, the ticks don't show up on the chart anymore. 
```
`
   const getTicks = (data) => {
		let ticks: any = []
		data.forEach(item => {
			const month = dayjs(item.valDate).get('M')
			const year = dayjs(item.valDate).get('y')

			if (!ticks.includes(`${month} ${year}`)) {
				ticks.push(`${month} ${year}`)
			}
		})

		// covert to ISO string since valDate is in ISO format
		ticks = ticks.map(item => dayjs(item, 'M YYYY').toISOString())

		/* console log
			[
    "2021-04-01T04:00:00.000Z",
    "2021-05-01T04:00:00.000Z",
    "2021-06-01T04:00:00.000Z",
    "2021-07-01T04:00:00.000Z",
    "2021-08-01T04:00:00.000Z",
    "2021-09-01T04:00:00.000Z"
]
		*/
	return ticks
}
`
```
my xAxis props:
```
xAxis: {
						dataKey: 'valDate',
						tickFormatter: (value: any) => dayjs(value).format("MMM'YY"),
						tickMargin: 5,
						ticks: getTicks(data),
					},
```

**Reply by shettyvega** [reply]
seems like the only way to do this is by converting date to unix format.

**Reply by ckifer** [reply]
by default XAxis is categorical. That means that each data point is an explicit "category" and has to exist in your dataset. Recharts cannot interpolate time with timestamps because it does not know that your categories are timestamps, only strings.

If you set `<XAxis type="number" />` and your data is numbers then recharts can interpolate between (since I know 2 is between 1 and 3). Eventually we will explicitly support time axes, but we can't at the moment. 

If you want these ticks at values that don't exist in your data, you'll need to use numeric data

**Reply by ckifer** [reply]
So in this case use `getTicks` to return unix time, and then `tickFormatter` to format to what you want to display to your users

**Reply by shettyvega** [reply]
Thank you. I could get it to work converting to unix format. Similarly, I have a repeating ticks issue on the y axis however it is in number format and straightforward data. Is using ticks array the solution here as well? 
```
					yAxis: { tickFormatter: (value: any) => formatPercent(value * 100, 0) },
```
I could get around doing this but it adds unequal gaps in ticks:
```
	yAxis: {
						domain: ['dataMin', 'dataMax'],
						interval: 'preserveStartEnd',
						tickFormatter: (value: any) => formatPercent(value * 100, 0),
						type: 'number',
					},
```
![image](https://github.com/recharts/recharts/assets/140361358/827aa163-5493-4851-b841-69f3947fbda7)
is there a better way to do this?

**Reply by ckifer** [reply]
Some options are:

* use `domain: 'auto'`
* Try `interval: 'equidistantPreserveStart'?,
* add another point of precision to your percentage formatter
* you can use the `ticks` array and make make it consist of n values based on the number of ticks you want to show

**Reply by shettyvega** [reply]
The last option seems to be the only clean one with my data. I can probably log a different issue request for this but why don't we have something similar to allowDuplicatedCategory for numbers? I would think it maybe easy to do?

**Reply by ckifer** [reply]
Because technically there are no duplicates. `allowDuplicatedCategory={false}` only prevents ticks from being created if they actually exist in your data, your formatter function is run way after axis information is generated and could still create them.

Same in this case, your data and where the ticks are being generated contain no duplicate numbers, those only exist after formatting is run for each datum.

If there were duplicates in your data a duplicate tick would not be generated

**Reply by shettyvega** [reply]
I misunderstood allowDuplicatedCategory in that case. Other than using ticks array, I cannot get a solution which ensures ticks are evenly spaced and has no duplicate ticks for number/time format. Thanks for your response.
